### ISSUE TITLE
Why not just put all of these into sklearn.datasets or move sklearn.datasets into this repo?

### ISSUE THREAD

**Issue by KOLANICH** [daviddiazvico/scikit-datasets#9]
It is a bit inconvenient to use tonn of packages.

So, if you are serious about creating a repo of code allowing importing datasets, `sklearn.datasets` should be moved here and this repo should be transfered to the org, and then we should do a lot of work about creating a fully machine-readable package of datasets.

If you are not serious, consider moving the code to sklearn.datasets.

**Comment by KOLANICH** [comment]
>They said that we should upload the datasets to OpenML instead.

OpenML is a good project (though buggy, I have tried to upload there some datasets and failed - the server returns a very strange error), but there are other repos of datasets. We need to do better.

1. currently we fetch datasets from every source with custom code. It'd be better to 
have these all in data, not code, and these data in a separate repo.

2. we want to have a metadata about the dataset, but some datasets lack them, and some store them in non-standardized way. 

3. we don't want to store datasets ourselves, so we don't want to be burdened with updating them.

>such as CRAN, are not in a standardized format, so they are not eligible to be added to Sklearn.

And we need to do something about it.

I have implemented a [fetcher of RDatasets](https://github.com/scikit-learn/scikit-learn/pull/12702), but it doesn't solve all the problems.

So the ideas:

1. create a specification of datasets metainformation format. https://frictionlessdata.io/specs/data-package/ should be helpful.

2. create the architecture. split fetching from parsing. implement several typical methods of retrieving column metadata, including a one of guessing it from natural language. If metadata cannot be fetched, we would store it in our repo.

3. create a repo and populate it with descriptions of datasets in the text format specified

4. transform the metadata from text format into records in sqlite db for rapid search when building the package and embed this db into wheel

**Comment by vnmabus** [comment]
> I have implemented a fetcher of RDatasets, but it doesn't solve all the problems.

Your fetcher fetches from a curated list of datasets, standardized to a CSV format. The fetcher provided here, fetches datasets from arbitrary R packages (if they are in the RData format). Neither CRAN nor R force developers to follow a particular structure in their datasets. Thus, it is impossible to (in general) translate them to the Sklearn format automatically.

Your ideas sound intriguing. I did not know about https://frictionlessdata.io/specs/data-package/ and I think it is worth a look.

**Comment by daviddiazvico** [comment]
I think moving this to scikit-learn-contrib as a separate project would the best option.

However, as Carlos just pointed out, the sklearn maintainers where not willing to integrate our fetchers into sklearn and I doubt they will accept that we take control of the sklearn.datasets code here.

In my opinion, it makes sense that the toy dataset loaders (iris, boston, etc., basically the ones that are so small that are included as files in the sklearn repo), the synthetic data generators (sklearn.datasets.samples_generator), and the datasets core (the Bunch class, some utility functions, etc.) are kept in sklearn. The small datasets are great to test new functionality in sklearn, are included in the repo, and do not define any validation partition, test partition, or CV folds.

I think only the big dataset fetchers (openml, kddcup, etc) should be moved to a separate dataset repo with our fetchers. In the first place because they have to be downloaded "on the fly", but also because big datasets usually pack a predefined validation or test partition, and in some cases even predefined CV folds, that have to be treated carefully with CV generators if one wants to make the experiments reproducible (take a look to my raetsch, libsvm, or keel fetchers for example).

What do you think? Unfortunately, right now I lack the time to discuss this with the people in charge of sklearn or sklearn-contrib, but I think this would be a nice initiative.

**Comment by KOLANICH** [comment]
>In my opinion, it makes sense that the toy dataset loaders (iris, boston, etc., basically the ones that are so small that are included as files in the sklearn repo), the synthetic data generators (sklearn.datasets.samples_generator), and the datasets core (the Bunch class, some utility functions, etc.) are kept in sklearn.

I agree. I have a PR into sklearn which may be useful for this repo: https://github.com/scikit-learn/scikit-learn/pull/12721

**Comment by daviddiazvico** [comment]
Sorry for the late response, I'm quite busy lately.

I've been thinking on this and I agree with you. I like the idea of decoupling sklearn algorithms from the data. Moreover, this could be a nice opportunity to align with [sklearn-pandas](https://github.com/scikit-learn-contrib/sklearn-pandas).
